how-we-work/how-to-upgrade-dependencies.md

```diff
@@ -0,0 +1,75 @@
+
+`npm audit` -- this command generates a report of known vulnerabilities in dependencies described in `cellxgene/client/package.json`.
+
+`npm outdated` -- this command generates a report of all dependencies in `cellxgene/client/package.json` and whether there are newer versions available.
+
+Run these two commands and upgrade packages to the latest available. Please also make any necessary changes in the client code to ensure that that the frontend still works as expected and is not using outdated references. Finally, run `npm i` to install the packages and update the corresponding `package-lock.json`.
+
+[Here](https://github.com/chanzuckerberg/cellxgene/pull/2167/files) is a sample PR that upgraded several frontend npm dependencies.
+
+### Upgrading cellxgene Explorer CZI-hosted backend
+
+The pip package dependency files for the CZI-hosted backend for cellxgene Explorer can be found at `cellxgene/backend/czi-hosted/requirements.txt` as well as `single-cell-infra/terraform/modules/hosted-cellxgene/common/customize/requirements.txt`. These two files should be in sync except that the file in `single-cell-infra` should pin all packages to a single version and the file in `cellxgene` should use the `>=` sign for the same version numbers. This is so that when we are testing locally, we can easily find out if a new release of a package is causing compatibility issues. We do want to pin versions in the `single-cell-infra` repo because these are the versions installed during deployment and we don't want an unexpected incompatibility due to a new package release to cause a production deployment to fail.
+
+To perform the upgrades you can make use of `pip-upgrader` to figure out which packages have recent updates:
+
+1. Install [pip-upgrader](https://pypi.org/project/pip-upgrader/): `pip install pip-upgrader`.
+
+2. Upgrade the packages based on the report and verify that the application still works (please fully deploy to a `dev` or `rdev` environment to double check everything).
+
+You can also make use of the [Snyk reports](https://app.snyk.io/org/cellxgene) to figure out which specific packages have security vulnerabilities as well as the [`safety` package](#Upgrade-cellxgene-Explorer-desktop-backend).
+
+Finally, besides the `pip` installed packages, please also take a look at the Dockerfile (in the top level directory `cellxgene/Dockerfile`) to ensure that there are no security vulnerabilities there. Snyk is your friend to figure out if there are and what the remediation process is.
+
+### Upgrading cellxgene Explorer desktop backend
+
+The policy we have to upgrade the backend of cellxgene Explorer _desktop_ is different than any of the other policies. Because many users may have different versions of the dependent `pip` packages already installed on their laptop, and because many users do not use a virtual environment, we need to be _as permissive as possible_ when defining what the allowable versions of each package are.
```

Comment:
This prose is somewhat unclear. Are we "permissive" as in allowing versions that have vulnerabilities?  Or as in allowing any version that is reasonably secure?

One possible policy:
* minimum version that is functionally compatible and includes no medium or high security risks
* it is acceptable to exclude specific versions (eg, `foo>=1.9.0,!=2.0.0`) if that is simpler

Reply:
Added more detailed prose. Your suggested policy is what I had in mind so pasted that in :)